### ISSUE TITLE
Centralise indented code block checks

### ISSUE THREAD

**Issue by chrisjsewell** [markdown-it/markdown-it#936]
In markdown-it-py (https://github.com/executablebooks/markdown-it-py/commit/9251695727cfa948bb18fc76a5dc85495cacc361) and for its plugins (https://github.com/executablebooks/mdit-py-plugins/commit/85f7ed312c530bb7faf98c079c71379122082556 + https://github.com/executablebooks/mdit-py-plugins/commit/9e5aff87a598504c46f9488a0d10423c698de9fa)
I have centralised the check - present in all block rules - for a block indented by greater than 4 spaces, i.e. the check for an indented code block, and deactivate it when the `code` rule is disabled.

This allows for behaviour similar to https://github.com/jgm/djot and https://mdxjs.com/, whereby you can indent syntax blocks arbitrarily, e.g.

```md
<div>
    <div>
        
        > arbitrarily indented syntax block

    </div>
</div>
```

at the moment, here, its a bit weird because most block rules will not work even if you turn off code blocks

**Comment by rlidwka** [comment]
Yeah, I had my doubts about this one as well for some time.

### Why?

Not clear what the behavior should be if code rule is disabled.

I wrote `markdown-it` rules to the letter of the spec, which says "must be no more than 4 spaces" on every block rule. But spec is of course written with the assumption that code block is always there.

`djot` is a great argument in favor of what you're suggesting. By the way, are there any other useful changes or commonmark deviations in there?

### How to check if code rule is present?

I don't know how you manage custom rules, but this looks a bit suspicious:

```py
# pre-check if code blocks are enabled, to speed up is_code_block method
self._code_enabled = "code" in self.md["block"].ruler.get_active_rules()
```

1. can user define their own block rule named `code`? what happens if they do?
2. what if user wants to define another block similar to `code`, but named differently, and still have the restriction?
3. and why is 4 spaces hardcoded?

### Just some thoughts

Maybe block state should have something like `max_starting_spaces` as a parameter, which should be `+inf` by default, and decreased to `4` if code rule is present *by the code rule itself*.

Maybe block state should also have `min_starting_spaces` as a parameter to clamp it from the other side. Those checks also happen somewhere.

TL;DR: I like the idea, unsure about implementation details.

**Comment by chrisjsewell** [comment]
> By the way, are there any other useful changes or commonmark deviations in there?

I'd recommend just reading https://github.com/jgm/djot#rationale, this essentially details the commonmark deviations and their rationale, which also derives from https://johnmacfarlane.net/beyond-markdown.html

I actually learnt of djot after asking about one of the changes it makes; to remove non-locality of reference link parsing: https://github.com/commonmark/commonmark-spec/issues/702

> I don't know how you manage custom rules, but this looks a bit suspicious

yeh certainly open to other ideas

> Maybe block state should have something like max_starting_spaces as a parameter, which should be +inf by default, and decreased to 4 if code rule is present *by the code rule itself*.

I like the idea, however, how could this work, given that individual rules don't have any "initialisation hook" themselves? (obviously, plugins can initialise things in `MarkdownIt.use`, but built-in rules have no such feature)

**Comment by chrisjsewell** [comment]
From https://github.com/rlidwka/markdown-it.rs/issues/7#issuecomment-1603117055, I see now why you suggest this 😄 

So I guess this is definitely possible to add in markdown-it.rs

**Comment by rlidwka** [comment]
> Yeh ughh, I'd be very wary of doing this as obviously any divergences make it more difficult to port fixes/improvements you make here and ensure markdown-it-py is kept in-sync

> I see now why you suggest this 😄

I've suggested this, because markdown-it right now is in de-facto feature freeze.

The project it was created for is up and running, so we don't need to fix anything. There's no time/money to make any major changes. And even if we did have resources, markdown-it is now used by too many people to justify any breaking changes.

If you are porting it to python however:
 - you don't have backward compatibility issues
 - you don't have to sync to upstream (because as I said feature-freeze)
 - you may have some time to do it

If you were to do that, I have some ideas how to improve the parser. Most of these you can see in my Rust port, but not all (for example, I'm thinking this entire thing should work under ECS paradigm, which probably helps with performance and usability). That's actually why I wrote the that thing - I found out that in Rust I can refactor things a lot faster than in JS, and quickly test any ideas.

**Comment by chrisjsewell** [comment]
> you don't have backward compatibility issues

well not on the same scale as here, but markdown-it-py has a fair few users, so certainly not trivial to change too much 😅 

But yes the Rust port is looking really cool thanks, so I would rather also invest my time there and seeing what I can do with https://github.com/chrisjsewell/markdown-it-pyrs

**Comment by rlidwka** [comment]
> so I would rather also invest my time there and seeing what I can do with

Not a discussion for this repo, but...

"Parser plugins cannot currently be written in Python and added dynamically to the parser."

What exactly is missing for that? I was thinking about reflection API already, and that's a good use-case for it.

**Comment by rlidwka** [comment]
Please send a pull request for this library.

We might get this change in (if performance impact is negligible).
